docs/source/cpp/streaming_execution.rst

```diff
@@ -175,9 +175,607 @@ their completion::
+            names.emplace_back(aug_name);
+            }
+        }
+
+        std::vector<arrow::compute::Expression> exprs;
+        for (const auto& name : names) {
+            exprs.push_back(arrow::compute::field_ref(name));
+        }
+
+        return arrow::compute::project(exprs, names);
+    }
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    options->projection = Materialize({});  // create empty projection
+
+    // construct the scan node
+    cp::ExecNode* scan;
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{
+                              /*dataset*/dataset, 
+                              /*scan_options*/options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            arrow::compute::MakeExecNode("scan", 
+                              plan.get(), 
+                              {}, 
+                              scan_node_options));
+
+``SinkNode``
+------------
+
+:class:`SinkNode` can be considered as the output or final node of an streaming 
+execution definition. :class:`arrow::compute::SinkNodeOptions` interface is used to pass 
+the required options. Requires 
+`std::function<arrow::Future<arrow::util::optional<arrow::compute::ExecBatch>>()>* generator`
+and `arrow::util::BackpressureOptions backpressure`. 
+
+Example::
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    arrow::compute::ExecNode* sink;
+
+    ARROW_ASSIGN_OR_RAISE(sink, arrow::compute::MakeExecNode("sink", plan.get(), {source},
+                                                  arrow::compute::SinkNodeOptions{&sink_gen}));
+
+
+The output can be obtained as a table::
+
+    // // // translate sink_gen (async) to sink_reader (sync)
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+      basic_data.schema, std::move(sink_gen), exec_context.memory_pool());
+
+    // // validate the ExecPlan
+    ABORT_ON_FAILURE(plan->Validate());
+    std::cout << "Exec Plan Created: " << plan->ToString() << std::endl;
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                        arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    std::cout << "Results : " << response_table->ToString() << std::endl;
+
+
+``ConsumingSinkNode``
+---------------------
+
+:class:`arrow::compute::ConsumingSinkNode` is a sink node that owns consuming the data and 
+will not finish until the consumption is finished.  Use SinkNode if you are
+transferring the ownership of the data to another system.  
+Use :class:`arrow::compute::ConsumingSinkNode` if the data is being consumed within the exec 
+plan (i.e. the exec plan should not complete until the consumption has completed).
```

Comment:
This is nearly tautological. I think the general purpose is that `ConsumingSinkNode` is used to consume the output within the exec plan, instead of getting a stream of batches to read.

Reply:
Yes, I missed this point. I went through the code and you're correct.